[50-FEAT] 응답 가능한 날짜 관련 추가 수정 및 타임 테이블 응답 형식 변경

binimini opened the pull request:
## 작업 목록
- [x] 타임 테이블 응답 형식 변경
- [x] 약속 제안 확정시 응답 가능한 날짜 검증 추가
- [x] Date 객체 UTC 형식 / util 및 DB KST 형식으로 통일
- [x] 약속제안 사용자 응답시 날짜 검증 수정
- [x] DATEONLY 경우도 통일된 형식으로 처리하도록 수정

## 세부 내용
- `timeUtil.checkTimeResponseList()` 내에서 `IndexOf()`로 Date 객체 자체를 비교하게 되면 
 객체의 동등성(객체가 가진 값의 일치)이 아닌 동일성(객체 저장 주소의 일치)을 비교해 
모두 응답가능하지 않은 날짜로 처리되어서 비교 방식을 수정했습니다. 
  ```ts
  async checkTimeResponseList(
      timeResponse: TimeRequest,
      minTime: Date,
      maxTime: Date,
      availDates: Date[]
    ) {
  //..
  if (!this.isPossibleDate(date, availDates) || timeList.length > count) return false;
  // ..
  }
  // 객체 == 비교 대신 내부 값 비교
  isSameDate(date: Date, other: Date) {
      return (
        other.getFullYear() == date.getFullYear() &&
        other.getMonth() == date.getMonth() &&
        other.getDate() == date.getDate()
      );
    },
  
    isPossibleDate(date: Date, candidates: Date[]) {
      return candidates.filter((candidate) => this.isSameDate(date, candidate)).length != 0;
    },
  ```

- 사용자가 약속 제안에 시간 응답 (time-response)하거나 약속 제안을 생성하는 경우 
  Event와 관련된 데이터는 안드로이드 단에서 필요가 없는 데이터라고 판단해 삭제했습니다. 
 (약속장 아닌 사용자가 시간 응답할 시 바디에 반환 데이터 없이 201로 처리되도록 수정했습니다.)

- 안드로이드단에서 공통적으로 Date로 처리하기 위해 promisingDate처럼 날짜만 의미있는(시간 무시) 경우에도 
 `yyyy-mm-ddThh:mm:ss` 형식으로 반환되도록 통일했습니다. 약속 제안 타임 테이블에서 컬러값도 number 타입으로 변경되었습니다.

- 약속 제안 응답된 타임 테이블 반환 형식 수정하고 약속 제안 확정시 검증 날짜와 시간 모두 확인하도록 수정했습니다.
 (약속 제안 타임 테이블 반환 형식 처리 코드가 깔끔하지 못한데 이후에 최대한 수정해보겠습니다. 😅 
  더 깔끔하게 짤 수 있는 방식 보이면 언제든지이야기해주세요!)

- 날짜를 동일한 타임존으로 비교하기 위해서 `new Date()`시에는 UTC / util 함수 비교시, DB, 반환시에는 KST로 사용하도록 설정했습니다.
 (~~배포 서버의 로컬 타임존도 KST로 변경해줘야할 것 같아요~~ 확인해보니 배포 서버 로컬 타임존도 Asia/Seoul이라  바로 테스트해도 될 것 같습니당~ 🙃)

## 논의 사항
- UTC와 KST 관련해서 고생을 좀 했는데 많이 헷갈리는 부분이라서 공유 드립니다!
[Date 관련 간략 기록](https://psychedelic-jackfruit-258.notion.site/JS-Date-c3f5693b319a4684b5f7522ada7690e6)

## 연결된 이슈
- #50

binimini committed:
feat: 시간 관련 유틸 함수 추가

- isSameDate(Date, Date):boolean 각 Date 객체 같은 날짜인지 비교
- getIndexFromMinTime(Date, Date, number):number minTime과 비교할 Date 받아서 unit 단위로 index 계산 후 반환

binimini committed:
feat: 반환 형식 수정

- date 형태로 묶어주기 위한 처리 추가
- getIndexFromMinTime() 올바르지 않은 비교 Date 들어올시 -1 처리

binimini committed:
feat: 시간 처리 형식 통일 및 타임 테이블 반환 형식 수정

- date만 사용하는 컬럼도 yyyy-mm-dd hh:mm:ss 형식으로 반환하도록 수정
- new Date()로 생성한 객체 자체는 UTC / util 함수, DB 저장은 KST
- date만 존재시에서 format팅시 KST로 변환해 09:00:00 되는 문제 처리 필요

binimini committed:
feat: 사용자 약속 제안 시간 응답 관련 수정

- PromisingDateService.findDatesById() promisingDate  반환시 Date 객체 배열로 반환하도록 수정
- 약속 제안 가능한 날짜 배열 반환시 yyyy-mm-dd hh:mm:ss 형식으로 통일되도록 수정
- timeUtil.checkTimeResponseList() 객체 비교 -> 날짜 및 배열 크기 비교로 수정

binimini committed:
refactor: Prettier 적용 및 콘솔 코드 삭제

binimini committed:
refactor: TimeTable 관련 인터페이스 분리

binimini committed:
feat: 약속 제안 확정시 검증 추가

- 약속 제안 응답 가능 날짜인지 검증 추가
- timeUtil.compareTime(Date, Date):number 두 Date의 시간(날짜 무시) 비교 첫번째 Date가 작을시 -1, 같을시 0, 클시 1 반환
- minTime, maxTime 날짜 무시 시간만 비교하도록 수정

cpplovelove commented on services/promising-service.ts:
```diff
@@ -143,26 +157,54 @@ class PromisingService {
+                  ...timeMap.get(dateStr)![key]!,
+                  new UserResponse(user)
+                ];
+              }
```
이 부분 timeMap을 인자로 가지는 함수로 분리해보면 좀 더 깔끔해지지 않을까요..?!

cpplovelove commented on utils/time.ts:
```diff
@@ -154,33 +154,26 @@ const timeUtil = {
-      if (timeList.length > count) {
-        return false;
-      }
+      if (!this.isPossibleDate(date, availDates) || timeList.length > count) return false;
```
요 부분ㅠㅠ 테스트해보면서 수정하다보니 완전 엉망으로 뒀네요.. 깔끔하게 고쳐주셔서 감사합니다..🙇‍♀️!!

cpplovelove commented on utils/time.ts:
```diff
@@ -79,12 +79,13 @@ const timeUtil = {
       let { startDate, endDate } = indexList[i];
       (startDate = startDate * time + minTimeDate), (endDate = endDate * time + minTimeDate);
 
-      let startHour = Math.trunc(startDate / 60) + 9,
```
야매로 계산하고 넘겨주고 있었는데... 수정해주셔서 감사합니다!!👰

cpplovelove reviewed (approved):
완전 완전 수고하셨습니다! 🙌
노션으로 남겨주신 UTC, KST 부분 잘 읽었습니다! time에서 로컬 타임 기준으로 조회하는 함수 기준으로 나눠서 조회해야 해야하는 부분을 알아봤어야 했는데 단순히 무지성 9시간을 더해서 계산한 제 자신을 캐삭하고 싶네요... 🥶😭 
약속 제안 타임 테이블 반환형식 코드 부분 필요한 부분들인 것 같아서 좀 더 깔끔하게 분리하려면 함수로 한번 더 쪼개주면 어떨까..? 생각해봤습니다!

binimini committed:
refactor: 함수 분리 및 변수 네이밍 개선

- bindUsersByDateTime() : 사용자 날짜/시간대에 따라 인덱스로 묶는 함수 분리
- transformEvents2MapAndUsers() : 쿼리 반환 데이터 날짜 문자열 (yyyy-mm-dd) key로 가지는 Map으로 변환 (전체 응답한 사용자 목록 생성)

binimini commented:
선영님 의견 참고해서 좀 더 가독성 있도록 수정했어용 감사합니당~ 머지할께요! 👍
